Readme.md
Simpel Accounting in python with persian date support and it is forked from http://www.freeamir.com/fa/


debian/changelog
amir (0.1) maverick; urgency=low

  * New release.

 -- hadi <hadi@jooyesh0>  Sun, 09 Jan 2011 16:04:18 +0330

amir (0.1) lucid; urgency=low

  * New release.

 -- hadi <hadi@Jooyesh00>  Sun, 02 May 2010 13:45:34 +0430

amir (0.1) lucid; urgency=low

  * Initial release.

 -- hadi <hadi@Jooyesh00>  Sun, 02 May 2010 09:58:42 +0430


amir/database.py
from sqlalchemy import create_engine
from sqlalchemy import Column, ForeignKey, ColumnDefault
from sqlalchemy import Integer, String, Date, Boolean, Unicode
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base

from migrate.versioning import api
from migrate import exceptions
# metadata = MetaData(bind=engine)
 
# create tables in database
# metadata.create_all
Base = declarative_base()
 
class Subject(Base):
    __tablename__ = "subject"
    id = Column(Integer, primary_key=True)
    code = Column(String(20), nullable=False)
    name = Column(Unicode(60), nullable=False)
    parent_id = Column(Integer, ColumnDefault(0), ForeignKey('subject.id'), nullable=False)
    lft = Column(Integer, nullable=False)
    rgt = Column(Integer, nullable=False)
    type = Column(Integer)      # 0 for Debtor, 1 for Creditor, 2 for both
    
    def __init__(self, code, name, parent_id, left, right, type):
        self.code = code
        self.name = name
        self.parent_id = parent_id
        self.lft = left
        self.rgt = right
        self.type = type

class Bill(Base):
    __tablename__ = "bill"
    id = Column(Integer, primary_key=True)
    number = Column(Integer, nullable = False)
    creation_date = Column(Date, nullable = False)
    lastedit_date = Column(Date, nullable = False)
    date = Column(Date, nullable = False)   #date of transactions in the bill
    permanent = Column(Boolean, ColumnDefault(False), nullable = False)
    
    def __init__(self, number, creation_date, lastedit_date, date, permanent):
        self.number = number
        self.creation_date = creation_date
        self.lastedit_date = lastedit_date
        self.date = date
        self.permanent = permanent  
    
class Notebook(Base):
    __tablename__ = "notebook"
    id = Column(Integer, primary_key=True)
    subject_id = Column(None, ForeignKey('subject.id'))
    bill_id = Column(None, ForeignKey('bill.id'))
    desc = Column(Unicode, ColumnDefault(""))
    value = Column(Integer, ColumnDefault(0), nullable = False)
    
    def __init__(self, subject_id, bill_id, value, desc):
        self.subject_id = subject_id
        self.bill_id = bill_id
        self.value = value
        self.desc = desc

class Database:
    def __init__(self, file, repository, echoresults):
        self.version = 1
        self.dbfile = file
        self.repository = repository
        
        #migrate code
        try:
            dbversion = api.db_version('sqlite:///%s' % file, self.repository)
        except exceptions.DatabaseNotControlledError:
            dbversion = 0
            api.version_control('sqlite:///%s' % file, self.repository, dbversion)
            
        if dbversion < self.version:
            api.upgrade('sqlite:///%s' % file, self.repository, self.version)
        
        engine = create_engine('sqlite:///%s' % file , echo=echoresults)
        
        metadata = Base.metadata
        metadata.create_all(engine)
    
        Session = sessionmaker(engine)
        self.session = Session()            
        
    def rebuild_nested_set(self, parent=0, left=0): 
        right = left+1;
        # get all children of this node  
        result = self.session.query(Subject.id).select_from(Subject).filter(Subject.parent_id == parent).all()
        for a in result :
            right = self.rebuild_nested_set(a[0], right);
 
        self.session.query(Subject).filter(Subject.id == parent).update(values = dict(lft = left,rgt = right))
        self.session.commit()
        
        return right+1;


amir/dateentry.py
import pygtk
import gtk
import gobject
import string
from datetime import date

import utility
from amirconfig import config
from calverter import calverter

def dateToString(date):
    if config.datetypes[config.datetype] == "jalali":
        jd = DateEntry.cal.gregorian_to_jd(date.year, date.month, date.day)
        (year, month, day) = DateEntry.cal.jd_to_jalali(jd)
    else:
        (year, month, day) = (date.year, date.month, date.day)
        
    datelist = ["", "", ""]
    datelist[config.datefields["year"]] = year
    datelist[config.datefields["month"]] = month
    datelist[config.datefields["day"]] = day
        
    delim = config.datedelims[config.datedelim]
    datestring = str(datelist[0]) + delim + str(datelist[1]) + delim + str(datelist[2])
    if config.digittype == 1:
        datestring = utility.convertToPersian(datestring)
    return datestring
    
class DateEntry(gtk.Entry):
    
    cal = calverter()
    
    def __init__(self, init_date=None):
        """
        date is a tuple containing the default DateEntry value. date is in the form of (YYYY, MM, DD) as three integers. 
        """
        gtk.Entry.__init__(self)
        self.set_alignment(0.5)
        self.connect("focus-out-event", self.correctDate)
#        self.connect("hide", self.correctDate)
        
        self.cal = calverter()

        if init_date != None:
            (self.year, self.month, self.day) = init_date
        else:
            today = date.today()
            if config.datetypes[config.datetype] == "jalali":
                jd = self.cal.gregorian_to_jd (today.year, today.month, today.day)
                jalali = self.cal.jd_to_jalali(jd)
                (self.year, self.month, self.day) = jalali
            else:
                self.year = today.year
                self.month = today.month
                self.day = today.day
                
        self.showDate(self.year, self.month, self.day)
        
    def showDate(self, year, month, day):
        datelist = ["", "", ""]
        datelist[config.datefields["year"]] = year
        datelist[config.datefields["month"]] = month
        datelist[config.datefields["day"]] = day
        
        delim = config.datedelims[config.datedelim]
        datestring = str(datelist[0]) + delim + str(datelist[1]) + delim + str(datelist[2])
        if config.digittype == 1:
            datestring = utility.convertToPersian(datestring)
        self.set_text(datestring)
        self.year = year
        self.month = month
        self.day = day
        
    #Assuming that date objects show gregorian date.
    def showDateObject(self, date):
        if config.datetypes[config.datetype] == "jalali":
            jd = self.cal.gregorian_to_jd(date.year, date.month, date.day)
            (jyear, jmonth, jday) = self.cal.jd_to_jalali(jd)
            self.showDate(jyear, jmonth, jday)
        else:
            self.showDate(date.year, date.month, date.day)
        
    def getDateObject(self):
        if config.datetypes[config.datetype] == "jalali":
            jd = self.cal.jalali_to_jd(self.year, self.month, self.day)
            (gyear, gmonth, gday) = self.cal.jd_to_gregorian(jd)
            return date(gyear, gmonth, gday)
        else :
            return date(self.year, self.month, self.day)
        
    def correctDate(self, sender, event):
        text = self.get_text()
        datelist = string.split(text, config.datedelims[config.datedelim]) 
        try:
            tyear = datelist[config.datefields["year"]]
            tyear = utility.convertToLatin(tyear)
        except IndexError:
            tyear = ""
        try:
            tmonth = datelist[config.datefields["month"]]
            tmonth = utility.convertToLatin(tmonth)
        except IndexError:
            tmonth = ""
        try:
            tday = datelist[config.datefields["day"]]
            tday = utility.convertToLatin(tday)
        except IndexError:
            tday = ""
        
        if config.datetypes[config.datetype] == "jalali":
            minyear = 1349
            baseyear = "1300"
        else:
            minyear = 1970
            baseyear = "2000"
            
        try:
            if len(tyear) > 4:
                year = self.year
            else:
                year = int(baseyear[:4-len(tyear)] + tyear)
                
            if year < minyear:
                year = minyear
        except ValueError:
            year = minyear
            
        try:
            month = int(tmonth)
            if month > 12:
                month = 12
            else:
                if month < 1:
                    month = 1
        except ValueError:
            month = 1
            
        try:
            day = int(tday)
            if day > 31:
                day = 31
            else:
                if day < 1:
                    day = 1
        except ValueError:
            day = 1
                
        if config.datetypes[config.datetype] == "jalali":
            jd = self.cal.jalali_to_jd(year, month, day)
            (gyear, gmonth, gday) = self.cal.jd_to_gregorian(jd)
        else:
            (gyear, gmonth, gday) = (year, month, day)
            
        correct = 0
        while correct == 0:
            try:
                testdate = date(gyear, gmonth, gday)
                correct = 1
            except ValueError:
                gday -= 1
                day -= 1
                
        self.showDate(year, month, day)
        self.year = year
        self.month = month
        self.day = day


amir/numberentry.py
import pygtk
import gtk
import gobject

import utility

class NumberEntry(gtk.Entry):
    
    def __init__(self):
        gtk.Entry.__init__(self)
        self.insert_sig = self.connect("insert-text", self.insert_cb)
    
    def insert(self, widget, text, pos):
    # the next three lines set up the text. this is done because we
    # can't use insert_text(): it always inserts at position zero.
        orig_text = unicode(widget.get_text())
        text = unicode(text)
        try:
            num = int(text)
        except ValueError:
            text = ""
            
        new_text = orig_text[:pos] + text + orig_text[pos:]
        
    # avoid recursive calls triggered by set_text
        widget.handler_block(self.insert_sig)
    # replace the text with some new text
        widget.set_text(new_text)
        widget.handler_unblock(self.insert_sig)
    # set the correct position in the widget
        widget.set_position(pos + len(text))
       
    def insert_cb(self, widget, text, length, position):
    # if you don't do this, garbage comes in with text
        text = text[:length]
        pos = widget.get_position()
    # stop default emission
        widget.emit_stop_by_name("insert_text")
        gobject.idle_add(self.insert, widget, text, pos)


amir/notebookreport.py
import pygtk
import gtk
from datetime import date
import os
import platform

from sqlalchemy import or_
from sqlalchemy.orm.util import outerjoin
from sqlalchemy.sql import between
from sqlalchemy.sql.functions import sum
from sqlalchemy.sql.operators import ColumnOperators

import numberentry
import subjects
import utility
import printreport
import previewreport
from database import *
from dateentry import *
from amirconfig import config
from helpers import get_builder

class NotebookReport:
    DAILY = 1
    LEDGER = 2
    SUBLEDGER = 3
    
    def __init__(self, type=1):
        self.builder = get_builder("report")
        
        self.window = self.builder.get_object("window1")
        
        self.code = numberentry.NumberEntry()
        box = self.builder.get_object("codebox")
        box.add(self.code)
        self.code.show()
        self.code.connect("activate", self.selectSubject)
        self.code.set_tooltip_text(_("Press Enter to see available subjects."))
        
        self.date = DateEntry()
        box = self.builder.get_object("datebox")
        box.add(self.date)
        self.date.set_sensitive(False)
        self.date.set_activates_default(True)
        self.date.show()
        
        self.fdate = DateEntry()
        box = self.builder.get_object("fdatebox")
        box.add(self.fdate)
        self.fdate.set_sensitive(False)
        self.fdate.set_activates_default(True)
        self.fdate.show()
        
        self.tdate = DateEntry()
        box = self.builder.get_object("tdatebox")
        box.add(self.tdate)
        self.tdate.set_sensitive(False)
        self.tdate.set_activates_default(True)
        self.tdate.show()
        
        self.fnum = numberentry.NumberEntry()
        box = self.builder.get_object("fnumbox")
        box.add(self.fnum)
        self.fnum.set_sensitive(False)
        self.fnum.set_activates_default(True)
        self.fnum.show()
        
        self.tnum = numberentry.NumberEntry()
        box = self.builder.get_object("tnumbox")
        box.add(self.tnum)
        self.tnum.set_sensitive(False)
        self.tnum.set_activates_default(True)
        self.tnum.show()
        
        self.builder.get_object("allcontent").set_active(True)
        
        self.type = type
        self.subcode = ""
        self.subname = ""
        
        self.window.show_all()
        self.builder.connect_signals(self)
        
        if type == self.__class__.DAILY:
            self.builder.get_object("subjectbox").hide()
            self.window.set_title(_("Daily NoteBook"))
        elif type == self.__class__.LEDGER:
            self.window.set_title(_("Ledgers NoteBook"))
        else:
            self.window.set_title(_("Sub-ledgers NoteBook"))
            
    def createReport(self):
        report_header = []
        report_data = []
        col_width = []
        remaining = 1
        query1 = config.db.session.query(Notebook, Subject.code, Bill)
        query1 = query1.select_from(outerjoin(outerjoin(Notebook, Subject, Notebook.subject_id == Subject.id), 
                                            Bill, Notebook.bill_id == Bill.id))
        query2 = config.db.session.query(sum(Notebook.value))
        query2 = query2.select_from(outerjoin(outerjoin(Notebook, Subject, Notebook.subject_id == Subject.id), 
                                            Bill, Notebook.bill_id == Bill.id))
        
        # Check if the subject code is valid in ledger and subledger reports
        if self.type != self.__class__.DAILY:
            code = utility.convertToLatin(self.code.get_text())
            query3 = config.db.session.query(Subject.name)
            query3 = query3.select_from(Subject).filter(Subject.code == code)
            names = query3.first()
            if names == None:
                errorstr = _("No subject is registered with the code: %s") % self.code.get_text()
                msgbox = gtk.MessageDialog(self.window, gtk.DIALOG_MODAL, gtk.MESSAGE_WARNING, gtk.BUTTONS_OK, errorstr)
                msgbox.set_title(_("No subjects found"))
                msgbox.run()
                msgbox.destroy()
                return
            else:
                self.subname = names[0]
                self.subcode = code
                query1 = query1.filter(Subject.code.startswith(code))
                query2 = query2.filter(Subject.code.startswith(code))
            
        searchkey = unicode(self.builder.get_object("searchentry").get_text())
        if searchkey != "":
            try:
                value = int(utility.convertToLatin(searchkey))
            except (UnicodeEncodeError, ValueError):  #search key is not a number
                query1 = query1.filter(Notebook.desc.like("%"+searchkey+"%"))
            else:        
                query1 = query1.filter(or_(Notebook.desc.like("%"+searchkey+"%"), Notebook.value == value, Notebook.value == -value))
        # Check the report parameters  
        if self.builder.get_object("allcontent").get_active() == True:
            query1 = query1.order_by(Bill.date.asc(), Bill.number.asc())
            remaining = 0
        else:
            if self.builder.get_object("atdate").get_active() == True:
                date = self.date.getDateObject()
                query1 = query1.filter(Bill.date == date).order_by(Bill.number.asc())
                query2 = query2.filter(Bill.date < date)
            else:
                if self.builder.get_object("betweendates").get_active() == True:
                    fdate = self.fdate.getDateObject()
                    tdate = self.tdate.getDateObject()
                    if tdate < fdate:
                        msgbox = gtk.MessageDialog(self.window, gtk.DIALOG_MODAL, gtk.MESSAGE_ERROR, gtk.BUTTONS_OK, 
                                                   _("Second date value shouldn't precede the first one."))
                        msgbox.set_title(_("Invalid date order"))
                        msgbox.run()
                        msgbox.destroy()
                        return
                    query1 = query1.filter(Bill.date.between(fdate, tdate)).order_by(Bill.date.asc(), Bill.number.asc())
                    query2 = query2.filter(Bill.date < fdate)
                else:
                    if unicode(self.fnum.get_text()) == '' or unicode(self.tnum.get_text()) == '':
                        msgbox = gtk.MessageDialog(self.window, gtk.DIALOG_MODAL, gtk.MESSAGE_ERROR, gtk.BUTTONS_OK, 
                                                   _("One of document numbers are empty."))
                        msgbox.set_title(_("Invalid document order"))
                        msgbox.run()
                        msgbox.destroy()
                        return
                    
                    fnumber = int(unicode(self.fnum.get_text()))
                    tnumber = int(unicode(self.tnum.get_text()))
                    if tnumber < fnumber:
                        msgbox = gtk.MessageDialog(self.window, gtk.DIALOG_MODAL, gtk.MESSAGE_ERROR, gtk.BUTTONS_OK, 
                                                   _("Second document number shouldn't be greater than the first one."))
                        msgbox.set_title(_("Invalid document order"))
                        msgbox.run()
                        msgbox.destroy()
                        return
                    query1 = query1.filter(Bill.number.between(fnumber, tnumber)).order_by(Bill.date.asc(), Bill.number.asc())
                    query2 = query2.filter(Bill.number < fnumber)
        
        #Prepare report data for PrintReport class
        res = query1.all()
        if self.type == self.__class__.DAILY:
            report_header = [_("Doc. Number"), _("Date"), _("Subject Code"), _("Description"), _("Debt"), _("Credit")]
            #define the percentage of table width that each column needs
            col_width = [10, 10, 11, 43, 13, 13 ]
            for n, code, b in res:
                desc = n.desc
                if n.value < 0:
                    credit = utility.showNumber("0")
                    debt = utility.showNumber(-(n.value))
                else:
                    credit = utility.showNumber(n.value)
                    debt = utility.showNumber("0")
                    desc = "   " + desc
                
                billnumber = str(b.number)   
                if config.digittype == 1:
                    code = utility.convertToPersian(code)
                    billnumber = utility.convertToPersian(billnumber)
                report_data.append((billnumber, dateToString(b.date), code, desc, debt, credit))
        else:
            diagnose = ""
            if remaining != 0:
                remaining = query2.first()[0]
            
            #if self.type == self.__class__.LEDGER:
            report_header = [_("Doc. Number"), _("Date"), _("Description"), _("Debt"), _("Credit"), _("Diagnosis"), _("Remaining")]
            #define the percentage of table width that each column needs
            col_width = [10, 10, 37, 13, 13, 4, 13]
            for n, code, b in res:
                if n.value < 0:
                    credit = utility.showNumber("0")
                    debt = utility.showNumber(-(n.value))
                else:
                    credit = utility.showNumber(n.value)
                    debt = utility.showNumber("0")
                    
                remaining += n.value
                billnumber = str(b.number)
                if config.digittype == 1:
                    billnumber = utility.convertToPersian(billnumber)
                if remaining < 0:
                    diagnose = _("deb")
                    report_data.append((billnumber, dateToString(b.date), n.desc, debt, credit, diagnose, utility.showNumber(-(remaining))))
                else:
                    if remaining == 0:
                        diagnose = _("equ")
                    else:
                        diagnose = _("cre")
                    report_data.append((billnumber, dateToString(b.date), n.desc, debt, credit, diagnose, utility.showNumber(remaining)))
    
#            else:
#                if self.type == self.__class__.SUBLEDGER:
#                    report_header = [_("Doc. Number"), _("Date"), _("Description"), _("Debt"), _("Credit"), _("Diagnosis"), _("Remaining")]
#                    col_width = [55, 64, 174, 70, 70, 20, 70]
#                    for n, code, b in res:
#                        if n.value < 0:
#                            credit = "0"
#                            debt = utility.showNumber(-(n.value))
#                        else:
#                            credit = utility.showNumber(n.value)
#                            debt = "0"
#
#                        remaining += n.value
#                        if remaining < 0:
#                            diagnose = _("deb")
#                            report_data.append((str(b.number), dateToString(b.date), n.desc, debt, credit, diagnose, utility.showNumber(-(remaining))))
#                        else:
#                            if remaining == 0:
#                                diagnose = _("equ")
#                            else:
#                                diagnose = _("cre")
#                            report_data.append((str(b.number), dateToString(b.date), n.desc, debt, credit, diagnose, utility.showNumber(remaining)))

        return {"data":report_data, "col-width":col_width ,"heading":report_header}
                
    def createPrintJob(self):
        report = self.createReport()
        if report == None:
            return
        if len(report["data"]) == 0:
            msgbox = gtk.MessageDialog(self.window, gtk.DIALOG_MODAL, gtk.MESSAGE_INFO, gtk.BUTTONS_OK, 
                                       _("The requested notebook is empty."))
            msgbox.set_title(_("Empty notebook"))
            msgbox.run()
            msgbox.destroy()
            return
        
        printjob = printreport.PrintReport(report["data"], report["col-width"], report["heading"])
        if self.type == self.__class__.DAILY:
            todaystr = dateToString(date.today())
            printjob.setHeader(_("Daily Notebook"), {_("Date"):todaystr})
            printjob.setDrawFunction("drawDailyNotebook")
        else:
            if config.digittype == 1:
                code = utility.convertToPersian(self.subcode)
            else:
                code = self.subcode
                
            if self.type == self.__class__.LEDGER:
                printjob.setHeader(_("Ledgers Notebook"), {_("Subject Name"):self.subname, _("Subject Code"):code})
            else:
                printjob.setHeader(_("Sub-ledgers Notebook"), {_("Subject Name"):self.subname, _("Subject Code"):code})
            printjob.setDrawFunction("drawSubjectNotebook")
        return printjob
            
    def createPreviewJob(self):
        report = self.createReport()
        if report == None:
            return
        if len(report["data"]) == 0:
            msgbox = gtk.MessageDialog(self.window, gtk.DIALOG_MODAL, gtk.MESSAGE_INFO, gtk.BUTTONS_OK, 
                                       _("The requested notebook is empty."))
            msgbox.set_title(_("Empty notebook"))
            msgbox.run()
            msgbox.destroy()
            return
            
        preview = previewreport.PreviewReport(report["data"], report["heading"])
        if self.type == self.__class__.DAILY:
            todaystr = dateToString(date.today())
            preview.setDrawFunction("drawDailyNotebook")
        else:
            if config.digittype == 1:
                code = utility.convertToPersian(self.subcode)
            else:
                code = self.subcode
                
            preview.setDrawFunction("drawSubjectNotebook")
        return preview
    
    def previewReport(self, sender):
        if platform.system() == 'Windows':
            printjob = self.createPreviewJob()
            if printjob != None:
                printjob.doPreviewJob()
        else:
            printjob = self.createPrintJob()
            if printjob != None:
                printjob.doPrintJob(gtk.PRINT_OPERATION_ACTION_PREVIEW)
    
    def printReport(self, sender):
        printjob = self.createPrintJob()
        if printjob != None:
            printjob.doPrintJob(gtk.PRINT_OPERATION_ACTION_PRINT_DIALOG)
        
    def exportToCSV(self, sender):
        report = self.createReport()
        if report == None:
            return
        
        content = ""
        for key in report["heading"]:
            content += key.replace(",", "") + ","
        content += "\n"
           
        for data in report["data"]:
            for item in data:
                content += item.replace(",", "") + ","
            content += "\n"
            
        dialog = gtk.FileChooserDialog(None, self.window, gtk.FILE_CHOOSER_ACTION_SAVE, (gtk.STOCK_CANCEL, gtk.RESPONSE_REJECT,
                                                                                         gtk.STOCK_SAVE, gtk.RESPONSE_ACCEPT))
        dialog.run()
        filename = os.path.splitext(dialog.get_filename())[0]
        file = open(filename + ".csv", "w")
        file.write(content)
        file.close()
        dialog.destroy()
    
    def selectSubject(self, sender):
        if self.type == self.__class__.LEDGER:
            subject_win = subjects.Subjects(True)
        else:
            subject_win = subjects.Subjects()
        code = self.code.get_text()
        subject_win.highlightSubject(code)
        subject_win.connect("subject-selected", self.subjectSelected)
    
    def subjectSelected(self, sender, id, code, name):
        if config.digittype == 1:
            code = utility.convertToPersian(code)
        self.code.set_text(code)
        sender.window.destroy()
        
    def atdate_toggled(self, sender):
        self.date.set_sensitive(sender.get_active())
        
    def betweendates_toggled(self, sender):
        active = sender.get_active()
        self.fdate.set_sensitive(active)
        self.tdate.set_sensitive(active)
        
    def betweendocs_toggled(self, sender):
        active = sender.get_active()
        self.fnum.set_sensitive(active)
        self.tnum.set_sensitive(active)
    


data/amir_migrate/versions/001_Version_1.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Copyright (C) 2010 <jooyeshgar> <info@jooyeshgar.com>
#This program is free software: you can redistribute it and/or modify it 
#under the terms of the GNU General Public License version 3, as published 
#by the Free Software Foundation.
#
#This program is distributed in the hope that it will be useful, but 
#WITHOUT ANY WARRANTY; without even the implied warranties of 
#MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR 
#PURPOSE.  See the GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License along 
#with this program.  If not, see <http://www.gnu.org/licenses/>.

from sqlalchemy import *
from migrate import *
import logging

meta = MetaData()

#New tables (version 1):
subject = Table('subject', meta,
    Column('id', Integer, primary_key=True),
    Column('code', String(20), nullable=False),
    Column('name', Unicode(60), nullable=False),
    Column('parent_id', Integer, ColumnDefault(0), ForeignKey('subject.id'), nullable=False),
    Column('lft', Integer, nullable=False),
    Column('rgt', Integer, nullable=False),
    Column('type', Integer)
)

bill = Table('bill', meta,
    Column('id', Integer, primary_key=True),
    Column('number', Integer, nullable = False),
    Column('creation_date', Date, nullable = False),
    Column('lastedit_date', Date, nullable = False),
    Column('date', Date, nullable = False),
    Column('permanent', Boolean, ColumnDefault(False), nullable = False)
)

notebook = Table('notebook', meta,
    Column('id', Integer, primary_key=True),
    Column('subject_id', Integer, ForeignKey('subject.id')),
    Column('bill_id', Integer, ForeignKey('bill.id')),
    Column('desc', Unicode, ColumnDefault("")),
    Column('value', Integer, ColumnDefault(0), nullable = False)
)

def upgrade(migrate_engine):
    # Upgrade operations go here. Don't create your own engine; bind migrate_engine
    # to your metadata
    meta.bind = migrate_engine
    subject.create(checkfirst=True)
    bill.create(checkfirst=True)
    notebook.create(checkfirst=True)
    
     #Insert test data
    op = subject.insert()
    op.execute({"id": 1, "code": "01", "name": u"بانکها", "parent_id": 0, "lft": 1, "rgt": 2, "type": 2},
               {"id": 2, "code": "02", "name": u"هزینه ها", "parent_id": 0, "lft": 3, "rgt": 28, "type": 0},
               {"id": 3, "code": "03", "name": u"موجودیهای نقدی", "parent_id": 0, "lft": 29, "rgt": 32, "type": 2},
               {"id": 4, "code": "04", "name": u"بدهکاران / بستانکاران", "parent_id": 0, "lft": 33, "rgt": 34, "type": 2},
               {"id": 5, "code": "06", "name": u"تراز افتتاحیه", "parent_id": 0, "lft": 35, "rgt": 38, "type": 2},
               {"id": 6, "code": "13", "name": u"اسناد دریافتی", "parent_id": 0, "lft": 39, "rgt": 42, "type": 2},
               {"id": 7, "code": "07", "name": u"موجودی اول دوره", "parent_id": 0, "lft": 43, "rgt": 46, "type": 2},
               {"id": 8, "code": "12", "name": u"جاری شرکا", "parent_id": 0, "lft": 47, "rgt": 50, "type": 2},
               {"id": 9, "code": "09", "name": u"دارایی ثابت", "parent_id": 0, "lft": 51, "rgt": 52, "type": 2},
               {"id": 10, "code": "0201", "name": u"حقوق پرسنل", "parent_id": 2, "lft": 4, "rgt": 5, "type": 0},
               {"id": 11, "code": "0202", "name": u"آب", "parent_id": 2, "lft": 6, "rgt": 7, "type": 0},
               {"id": 12, "code": "0203", "name": u"برق", "parent_id": 2, "lft": 8, "rgt": 9, "type": 0},
               {"id": 13, "code": "0204", "name": u"تلفن", "parent_id": 2, "lft": 10, "rgt": 11, "type": 0},
               {"id": 14, "code": "0301", "name": u"صندوق", "parent_id": 3, "lft": 30, "rgt": 31, "type": 2},
               {"id": 15, "code": "0601", "name": u"-", "parent_id": 5, "lft": 36, "rgt": 37, "type": 2},
               {"id": 16, "code": "0701", "name": u"-", "parent_id": 7, "lft": 44, "rgt": 45, "type": 2},
               {"id": 17, "code": "10", "name": u"خرید", "parent_id": 0, "lft": 53, "rgt": 56, "type": 0},
               {"id": 18, "code": "11", "name": u"فروش", "parent_id": 0, "lft": 57, "rgt": 60, "type": 1},
               {"id": 19, "code": "1001", "name": u"-", "parent_id": 17, "lft": 54, "rgt": 55, "type": 0},
               {"id": 20, "code": "1101", "name": u"-", "parent_id": 18, "lft": 58, "rgt": 59, "type": 1},
               {"id": 21, "code": "08", "name": u"سرمایه اول دوره", "parent_id": 0, "lft": 61, "rgt": 64, "type": 2},
               {"id": 22, "code": "14", "name": u"اسناد پرداختی", "parent_id": 0, "lft": 65, "rgt": 68, "type": 2},
               {"id": 23, "code": "05", "name": u"درآمدهای متفرقه", "parent_id": 0, "lft": 69, "rgt": 76, "type": 1},
               {"id": 24, "code": "15", "name": u"تخفیفات نقدی خرید", "parent_id": 0, "lft": 77, "rgt": 80, "type": 2},
               {"id": 25, "code": "16", "name": u"تخفیفات نقدی فروش", "parent_id": 0, "lft": 81, "rgt": 84, "type": 2},
               {"id": 26, "code": "0205", "name": u"گاز", "parent_id": 2, "lft": 12, "rgt": 13, "type": 0},
               {"id": 27, "code": "0206", "name": u"تخفیفات - تعدیل حساب", "parent_id": 2, "lft": 14, "rgt": 15, "type": 0},
               {"id": 28, "code": "0207", "name": u"حمل", "parent_id": 2, "lft": 16, "rgt": 17, "type": 0},
               {"id": 29, "code": "0208", "name": u"ضایعات کالا", "parent_id": 2, "lft": 18, "rgt": 19, "type": 0},
               {"id": 30, "code": "0209", "name": u"عوارض شهرداری", "parent_id": 2, "lft": 20, "rgt": 21, "type": 0},
               {"id": 31, "code": "0210", "name": u"کارمزد بانک", "parent_id": 2, "lft": 22, "rgt": 23, "type": 0},
               {"id": 32, "code": "0211", "name": u"کرایه", "parent_id": 2, "lft": 24, "rgt": 25, "type": 0},
               {"id": 33, "code": "0212", "name": u"مالیات", "parent_id": 2, "lft": 26, "rgt": 27, "type": 0},
               {"id": 34, "code": "0501", "name": u"دریافت تخفیف بابت تعدیل حساب", "parent_id": 23, "lft": 70, "rgt": 71, "type": 1},
               {"id": 35, "code": "0502", "name": u"حق‌العمل", "parent_id": 23, "lft": 72, "rgt": 73, "type": 1},
               {"id": 36, "code": "0503", "name": u"متفرقه", "parent_id": 23, "lft": 74, "rgt": 75, "type": 1},
               {"id": 37, "code": "1201", "name": u"جاری شرکا", "parent_id": 8, "lft": 48, "rgt": 49, "type": 1},
               {"id": 38, "code": "17", "name": u"برگشت از خرید", "parent_id": 0, "lft": 85, "rgt": 88, "type": 2},
               {"id": 39, "code": "18", "name": u"برگشت از فروش", "parent_id": 0, "lft": 89, "rgt": 92, "type": 2},
               {"id": 40, "code": "1701", "name": u"-", "parent_id": 38, "lft": 86, "rgt": 87, "type": 2},
               {"id": 41, "code": "1801", "name": u"-", "parent_id": 39, "lft": 90, "rgt": 91, "type": 2},
               {"id": 42, "code": "1501", "name": u"-", "parent_id": 24, "lft": 78, "rgt": 79, "type": 2},
               {"id": 43, "code": "1601", "name": u"-", "parent_id": 25, "lft": 82, "rgt": 83, "type": 2},
               {"id": 44, "code": "1301", "name": u"-", "parent_id": 6, "lft": 40, "rgt": 41, "type": 2},
               {"id": 45, "code": "0801", "name": u"-", "parent_id": 21, "lft": 62, "rgt": 63, "type": 2},
               {"id": 46, "code": "1401", "name": u"-", "parent_id": 22, "lft": 66, "rgt": 67, "type": 2})
    
    logging.debug("upgrade to 1")
     

def downgrade(migrate_engine):
    # Operations to reverse the above upgrade go here.
    logging.debug("downgrade to 0")
